Bayesian easy

marcpare opened the pull request:
Attacks the 5th percentile score by targeting "easy" words -- those that are possible to guess with 2 or 3 guesses using optimal strategy.

Achieves epsilon = 21.0 in 10,001 trials reliably (well, it has at least in the three times I've run it so far)

Easy words pulled from this strategy file:
https://sonorouschocolate.com/notes/images/c/c4/Optimaltree.normalmode.txt

Any leaf node with GGGGG2 or GGGGG3 is considered "easy"

From there, I update the Bayesian solver to only guess from the "easy" list after exactly 2 guesses with epsilon = 10.5

Definitely some more room to optimize from here; I chose 10.5 to come in comfortably under the 5% target. I bet the word list can be further refined as well.

marcpare commented:
Ok, pared down this idea significantly into another strategy ("Birdie") that gets total epsilon = 19.0 for the 5th percentile reliably.

The guess of "trace" solves Wordle in 2 guess 6.47% of the time
https://aprllp.com/wordle-how-to-maximise-your-chances-of-getting-it-in-two/

So, we allocate all the privacy budget to this first guess, then make a best effort on the second guess.

Return a word that is 100% consistent with the clue if possible, otherwise the word with the least number of inconsistencies.

This idea also works for the 50th% percentile case. Using a strategy that maximizes ["win in three"](https://drive.google.com/file/d/1oSDhy8GVw9h6CLeEDgrxNOM-LQsNuVsy/view) I was able to get a total epsilon = 50. Code is a mess; if I have a minute later I'll add it!

TedTed commented on easy_words.txt:
Can you move this to `strategies/`?

TedTed commented on evaluate.py:
When you push this, please revert back to 1001 for NUM_TRIALS, and put the 3 winning strategies in `STRATEGIES_UNDER_TEST`

TedTed commented on strategies/bayesian_easy.py:
If birdie works better, you can remove this (and related codes/files) from the PR before pushing it =)

TedTed reviewed (changes requested):
Very nice! I love how simple & effective Birdie is for the 5th percentile 💙

Some preliminary feedback on the PR below =)

marcpare commented:
Glad you liked it! I renamed the idea to "TwoGuess" and cleaned out the strictly worse "BayesianEasy" strategy.

I also added a little demo of exhaustively searching the game space to compute the exact likelihood of success.

In another script, I ran this for a range of epsilons, and you can see the nice shape of the exponential mechanism:

![birdie_plot](https://github.com/user-attachments/assets/2180952b-e105-401f-82ab-e213fe8a9ec3)

The exact epsilon for succeeding 5% of the time this strategy is ~17.85

Oh, and I wrote another script that exhaustively searches for the best starting word. Even with the slight change to the rules this repo implements for wordl, it still comes out to the same results I found by google search:

    trace 6.48
    crate 6.39
    salet 6.39
    reast 6.35
    slate 6.35
    carte 6.31
    parse 6.31
    caret 6.26
    peart 6.26
    carle 6.22

TedTed commented:
Merged! I rounded up to 17.9 for consistency with other scores. Super neat graph and strategy. I wonder if it's even possible to do better…
